swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,318 @@
+      auto moduleTriple = swift::getTargetSpecificModuleTriple(triple);
+      aliases.push_back(location + moduleTriple.normalize() + ".swiftmodule");
+    }
+  }
+
+  return aliases;
+}
+
+namespace codeql {
+
+std::unordered_map<std::string, std::string> rewriteOutputsInPlace(
+    SwiftExtractorConfiguration& config,
+    std::vector<std::string>& CLIArgs) {
+  std::unordered_map<std::string, std::string> remapping;
+
+  // TODO: handle filelists?
+  std::unordered_set<std::string> pathRewriteOptions({
+      "-emit-dependencies-path",
+      "-emit-module-path",
+      "-emit-module-doc-path",
+      "-emit-module-source-info-path",
+      "-emit-objc-header-path",
+      "-emit-reference-dependencies-path",
+      "-index-store-path",
+      "-module-cache-path",
+      "-o",
+      "-pch-output-dir",
+      "-serialize-diagnostics-path",
+  });
+
+  std::unordered_set<std::string> outputFileMaps(
+      {"-supplementary-output-file-map", "-output-file-map"});
+
+  std::vector<size_t> outputFileMapIndexes;
+  std::vector<std::string> maybeInput;
+  std::string targetTriple;
+
+  std::vector<std::string> newLocations;
+  for (size_t i = 0; i < CLIArgs.size(); i++) {
+    if (pathRewriteOptions.count(CLIArgs[i])) {
+      auto oldPath = CLIArgs[i + 1];
+      auto newPath = config.tempArtifactDir + '/' + oldPath;
+      CLIArgs[++i] = newPath;
+      newLocations.push_back(newPath);
+
+      remapping[oldPath] = newPath;
+    } else if (outputFileMaps.count(CLIArgs[i])) {
+      // collect output map indexes for further rewriting and skip the following argument
+      // We don't patch the map in place as we need to collect all the input files first
+      outputFileMapIndexes.push_back(++i);
+    } else if (CLIArgs[i] == "-target") {
+      targetTriple = CLIArgs[++i];
+    } else if (CLIArgs[i][0] != '-') {
+      // TODO: add support for input file lists?
+      // We need to collect input file names to later use them to extract information from the
+      // output file maps.
+      maybeInput.push_back(CLIArgs[i]);
+    }
+  }
+
+  for (auto index : outputFileMapIndexes) {
+    auto oldPath = CLIArgs[index];
+    auto maybeNewPath = rewriteOutputFileMap(config, oldPath, maybeInput, remapping);
+    if (maybeNewPath) {
+      auto newPath = maybeNewPath.value();
+      CLIArgs[index] = newPath;
+      remapping[oldPath] = newPath;
+    }
+  }
+
+  // This doesn't really belong here, but we've got Xcode...
+  for (auto& [oldPath, newPath] : remapping) {
+    llvm::StringRef path(oldPath);
+    auto aliases = computeModuleAliases(path, targetTriple);
+    for (auto& alias : aliases) {
+      remapping[alias] = newPath;
+    }
+  }
+
+  return remapping;
+}
+
+void ensureNewPathsExist(const std::unordered_map<std::string, std::string>& remapping) {
+  for (auto& [_, newPath] : remapping) {
+    llvm::SmallString<PATH_MAX> filepath(newPath);
+    llvm::StringRef parent = llvm::sys::path::parent_path(filepath);
```

Comment:
It could be a directory as well, yes.
So far I only observed that output files are not created if the parent directory does not exist. The directories seem to be also created, but that requires more testing on real world projects.